packages/core/src/fhir-deduplication/deduplicate-fhir.ts

```diff
@@ -29,350 +30,285 @@ import { createRef } from "./shared";
+    ...immunizationsResult.danglingReferences,
+    ...proceduresResult.danglingReferences,
+    ...obsSocialResult.danglingReferences,
+    ...obsLabsResult.danglingReferences,
+    ...obsVitalsResult.danglingReferences,
+    ...obsOthersResult.danglingReferences,
+    ...locationsResult.danglingReferences,
+    ...organizationsResult.danglingReferences,
+    ...relatedPersonsResult.danglingReferences,
+    ...famMemHistoriesResult.danglingReferences,
+    ...coveragesResult.danglingReferences,
+  ]);
+
+  // Combine all the remaining replacementMaps into one map
+  const combinedReplacementMap = new Map<string, string>([
+    ...medAdminsResult.refReplacementMap,
+    ...medRequestResult.refReplacementMap,
+    ...medStatementResult.refReplacementMap,
+    ...practitionersResult.refReplacementMap,
+    ...conditionsResult.refReplacementMap,
+    ...allergiesResult.refReplacementMap,
+    ...encountersResult.refReplacementMap,
+    ...diagReportsResult.refReplacementMap,
+    ...immunizationsResult.refReplacementMap,
+    ...proceduresResult.refReplacementMap,
+    ...obsSocialResult.refReplacementMap,
+    ...obsLabsResult.refReplacementMap,
+    ...obsVitalsResult.refReplacementMap,
+    ...obsOthersResult.refReplacementMap,
+    ...locationsResult.refReplacementMap,
+    ...organizationsResult.refReplacementMap,
+    ...relatedPersonsResult.refReplacementMap,
+    ...famMemHistoriesResult.refReplacementMap,
+    ...coveragesResult.refReplacementMap,
+  ]);
+
+  resourceArrays = replaceResourceReferences(resourceArrays, combinedReplacementMap);
+
+  // Remove resources with dangling links
+  const { updatedResourceArrays, deletedRefs } = removeResourcesWithDanglingLinks(
+    resourceArrays,
     danglingLinks
   );
+  resourceArrays = updatedResourceArrays;
 
-  const compositionsNoDangling = removeResourcesWithDanglingLinks(
-    resourceArrays.compositions,
-    danglingLinks
+  const { updatedResourceArrays: updatedResourceArrays2 } = removeResourcesWithDanglingLinks(
+    resourceArrays,
+    deletedRefs
   );
+  resourceArrays = updatedResourceArrays2;
 
-  const deduplicatedBundle: Bundle = cloneDeep(fhirBundle);
-  deduplicatedBundle.entry = [...deduplicatedNoDangling, ...compositionsNoDangling].map(
-    buildBundleEntry
-  );
+  const deduplicatedEntries: BundleEntry<Resource>[] = [];
+  // Rebuild the entries with deduplicated resources and add whatever is left unprocessed
+  for (const [, resources] of Object.entries(resourceArrays)) {
+    const entriesArray = resources && Array.isArray(resources) ? resources : [resources];
+    const entriesFlat = entriesArray.flatMap(v => v || []);
+    const entriesWithDeduplicatedReferences = entriesFlat.map(removeDuplicateReferences);
+    deduplicatedEntries.push(...entriesWithDeduplicatedReferences);
+  }
 
-  deduplicatedBundle.total = deduplicatedNoDangling.length;
+  // Rebuild the final bundle
+  const deduplicatedBundle: Bundle = cloneDeep(fhirBundle);
```

Comment:
can we put this line at the top and use the copy for everything here? just wanna avoid any risk of alternating the original bundle